packages/firestore/src/remote/bloom_filter.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Md5, Integer } from '@firebase/webchannel-wrapper';
+
+import { debugAssert } from '../util/assert';
+
+export class BloomFilter {
+  private readonly bitSize: number;
+
+  constructor(
+    private readonly bitmap: Uint8Array,
+    padding: number,
+    private readonly hashCount: number
+  ) {
+    this.bitSize = this.bitmap.length * 8 - padding;
```

Comment:
Please move the assertions to the very top of the constructor, before any other work. The assertions should be performed on the arguments, not the values calculated from the arguments. Also, it is _not_ required for `hashCount` to be zero when the bitmap is empty: an empty bloom filter can still have a non-zero number of hash functions. HOWEVER, an empty bitmap _must_ have padding==0, so that is a valid thing to assert.